File: 0901-1000/0947-Most Stones Removed with Same Row or Column/0947-Most Stones Removed with Same Row or Column.py
class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        parent = {}
        island = 0
        def findParent(x):
            if x not in parent:
                parent[x] = x
                nonlocal island
                island += 1
            else:
                while parent[x] != x:
                    parent[x] = parent[parent[x]]
                    x = parent[x]
            return x
        for row, col in stones:
            pr = findParent(row)
            pc = findParent(~col)
            if pr != pc:
                parent[pr] = pc
                island -= 1
        return len(stones) - island

File: 0901-1000/0947-Most Stones Removed with Same Row or Column/0947-Most Stones Removed with Same Row or Column.cpp
class Solution {
public:
    int removeStones(vector<vector<int>>& stones) {
        int island = 0;
        for (auto stone : stones) {
            int pr = findParent(stone[0], island);
            int pc = findParent(~stone[1], island);
            if (pr != pc) {
                parent[pr] = pc;
                --island;
            }
        }

        return stones.size() - island;
    }
    
private:
    unordered_map<int, int> parent;
    int findParent(int x, int& island) {
        if (!parent.count(x)) {
            parent[x] = x;
            ++island;
        }
        else {
            while (parent[x] != x) {
                parent[x] = parent[parent[x]];
                x = parent[x];
            }            
        }
        return x;
    }
};
